[Image Uploads Improvement] Add entry points to the error screen

hafizrahman opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #13419
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
Adds a new permanent CTA when product image uploads fail, so that user can access an error screen to see the details of the failed upload an retry. The behavior of this new CTA is as follows: 
- It will be shown as soon as the first image upload error happens
- It will stay visible until "Add photos is tapped again" for the given product. Or when navigating back from product detail with unsaved changes and clicking "Discard" changes from the confirm exit dialog CTA. 
- Currently, when the app is killed all upload errors are cleared (because we keep them in memory). This behavior probably makes sense since killing the app will also cancel the ongoing uploads. 
- A snack bar is shown for a few seconds for every upload error that happens. We might want to get rid of this snackbar in iteration II completely. Something to discuss with Ann. But for now I think it provides better visibility for each time an upload fails.

### Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->
Before running the testing steps, enable API Faker to make upload media requests. Apply the following changes from More menu > Settings > Developer Options > API faker: 

<img width="300"  src="https://github.com/user-attachments/assets/0fcd3d74-65de-4e39-aac5-9c3ec06c285d"> 


### Testing information

With API faker is enabled: 
1. Open product detail
2. Open the product images screen
3. Tap `Add photos`
4. Tap `Choose from device`
5. Select a photo and let the uploads fail.
6. See how a snackbar with the number of failed uploads is shown for a couple of seconds
7. See the new CTA "View recent upload failures" is shown
8. Tap on it, and enter error screen. Navigate back. 
9. Navigate back to product detail and check the new "View recent upload failures" is shown
10. Tap on it and click retry. 
11. See how you are taken back to the previous screen while the retry happens 
12. After getting some upload errors, edit some data from the product detail, then navigate back, tap on "DIscard" cta from confirmation dialog. Once landed on the product list, reopen the same product detail and check the button "View recent upload failures" is gone. 

### The tests that have been performed
<!-- To give the reviewer an idea of what could be missed in terms of testing -->
The above 

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->

https://github.com/user-attachments/assets/d8ec43e7-ee9b-4b89-b29a-e70d257c8377

- [X] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on big (tablet) and small (phone) in case of UI changes, and no regressions are added.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

hafizrahman committed:
Add button to view recent upload failure on product images screen

hafizrahman committed:
Observe current upload errors and hide/show button based on it.

hafizrahman committed:
Simplify observe logic

hafizrahman committed:
Make button navigate to upload screen

hafizrahman committed:
Add button to go to upload error screen when there's error on product details

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

JorgeMucientes committed:
Do not clear all image errors when entering error upload image screen

JorgeMucientes committed:
Merge branch 'trunk' into issue/13370-pt-ii-entry-points-to-error-screen

JorgeMucientes committed:
Add function to clear upload error from list, upon retrying

JorgeMucientes committed:
Remove unnecessary comment

JorgeMucientes committed:
Replace CTA from upload error snackbar

We now display a permanent access point to open upload errors screen

JorgeMucientes committed:
Set red text to highlight is error screen

This will highlight the CTA after the user experiences some upload error

JorgeMucientes committed:
Update upload error snackbar from product detail screen as well

JorgeMucientes committed:
Fix detekt identation issues

JorgeMucientes committed:
Fix and add unit tests

JorgeMucientes committed:
Clear image upload errors when exiting product details discarding changes

JorgeMucientes commented:
Hey @hafizrahman I've tweaked your implementation a bit. I'd appreciate if you could test out the changes and let me know what you think 🙏🏼

hafizrahman committed:
Update button placement on product details.

hafizrahman commented:
@JorgeMucientes The testing description point 12 says `"Once it fails again, close product detail and reopen it. The "View recent upload failures" should be gone"`. But from the PR description and code, it will only be gone if there's unsaved changes/current upload then merchants agree to go back.

Otherwise, when going back from details to list when there's upload error but no new changes, then back to details again, the error will persist. Like in video below:


https://github.com/user-attachments/assets/c5034ccb-9ba5-47b8-8a29-20be8fbbf246

In other words, I think the code is correct, but the testing description need updating. Does that sound right?

---

Overall though I think the code looks good, thanks for adding the tweaks with handling / clearing up errors. I added a commit https://github.com/woocommerce/woocommerce-android/pull/13389/commits/d7c0eef50405f76ff2a273d432f65143aa880740 to tweak button placement on product details, so they look like this now: 

| Dark | Light |
|-|-|
| ![Screenshot_1738290727](https://github.com/user-attachments/assets/392d7777-a18f-4fe8-9d66-f34034fd6a54) | ![Screenshot_1738290740](https://github.com/user-attachments/assets/6f06e83e-2cce-483d-b23a-9bb496cd5b52) |


I think for iteration 2 we can look at improving the button design if needed, but usability-wise I think it's OK for now.

JorgeMucientes commented:
Thanks for the careful review @hafizrahman. 

> In other words, I think the code is correct, but the testing description needs updating. Does that sound right?
 You're right. I've updated the instructions accordingly. 

Also, agree with the added UI change to the upload error CTA 👍🏼

JorgeMucientes committed:
Extract clearing upload error to view model

enqueueUpload() function should not be responsible for clearing past errors from the upload status.

JorgeMucientes committed:
Merge branch 'issue/13370-pt-ii-entry-points-to-error-screen' of https://github.com/woocommerce/woocommerce-android into issue/13370-pt-ii-entry-points-to-error-screen
